### ISSUE TITLE
Grpc.Tools: Using `OutputDir` and `GrpcOutputDir` in a '`.csproj`'-file leads to a build break

### ISSUE THREAD

**Issue by macrogreg** [grpc/grpc#29161]
This issue is similar to https://github.com/grpc/grpc/issues/19816
However, that is marked as resolved, but it re-occurs for me. Also, the warning causes a subsequent build failure.

### What version of gRPC and what language are you using?

* "Grpc" Version="2.44.0"
* "Grpc.Tools" Version="2.44.0"
* "Google.Protobuf" Version="3.19.4"

### What operating system (Linux, Windows,...) and version?

Windows 10

### What runtime / compiler are you using (e.g. python version or version of gcc)

Visual Studio 2022

### What did you do?

Tried to compile a "gRPC Client Project" by clicking Build from within VS.
Full repro below.

### What did you expect to see?

Project should build successfully.

### What did you see instead?

Build errors & warnings.

### Detailed info

This is the project file that fails:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <ItemGroup>    
    <PackageReference Include="Grpc" Version="2.44.0" />
    <PackageReference Include="Grpc.Tools" Version="2.44.0" PrivateAssets="All" />
    <PackageReference Include="Google.Protobuf" Version="3.19.4" />
  </ItemGroup>

  <PropertyGroup>    
    <ProtoApiRepoBaseDir>$(EnlistmentRoot)\..\api</ProtoApiRepoBaseDir>
    <AutoGenOutputBaseDir>$(MSBuildThisFileDirectory)AutoGenOutput</AutoGenOutputBaseDir>
  </PropertyGroup>

  <ItemGroup>

  <Protobuf Include = "$(ProtoApiRepoBaseDir)\**\*.proto" 
            Link = "ProtoApi\%(RecursiveDir)%(Filename)%(Extension)" 
            Access = "Public"
            ProtoCompile = "True"
            ProtoRoot = "$(ProtoApiRepoBaseDir)"
            CompileOutputs = "true"
            OutputDir = "$(AutoGenOutputBaseDir)"
            GrpcOutputDir = "$(AutoGenOutputBaseDir)"         
            GrpcServices = "client" />
  </ItemGroup>
  
</Project>
```

This leads initially a few build warnings. Here is one for example:

```
Message: Some expected protoc outputs were not generated.
File: Google.Protobuf.Tools.targets
Line: 330

    c:\Repos\SDK.NET\sdk-dotnet\Src\SDK\RawClient-Managed\AutoGenOutput\dependencies\gogoproto\GogoGrpc.cs
    c:\Repos\SDK.NET\sdk-dotnet\Src\SDK\RawClient-Managed\AutoGenOutput\temporal\api\command\v1\MessageGrpc.cs    
   [... a total of 24 XxxGrpc.cs files ...]
    c:\Repos\SDK.NET\sdk-dotnet\Src\SDK\RawClient-Managed\AutoGenOutput\temporal\api\workflow\v1\MessageGrpc.cs
```

The build process continues, but then fails with build errors like this one:

```
Error CS2001
Source file 'c:\Repos\SDK.NET\sdk-dotnet\Src\SDK\RawClient-Managed\AutoGenOutput\temporal\api\enums\v1\ResetGrpc.cs'
could not be found.
Project: Temporal.GrpcClient.Managed
```

There are 24 of these errors, one for each missing file.
This happens for each of the 3 target Fx versions I am using:
* Project `Temporal.GrpcClient.Managed (net462)`
* Project `Temporal.GrpcClient.Managed (netcoreapp3.1)`
* Project `Temporal.GrpcClient.Managed (net6.0)`

Indeed, those files are not in the output directory:

```
 Directory of c:\Repos\SDK.NET\sdk-dotnet\Src\SDK\RawClient-Managed\AutoGenOutput\temporal\api\enums\v1

2022-03-18  15:10             4,306 CommandType.cs
2022-03-18  15:10             3,438 Common.cs
2022-03-18  15:10            15,804 EventType.cs
2022-03-18  15:10            11,366 FailedCause.cs
2022-03-18  15:10             3,193 Namespace.cs
2022-03-18  15:10             3,153 Query.cs
2022-03-18  15:10             2,130 Reset.cs
2022-03-18  15:10             3,579 TaskQueue.cs
2022-03-18  15:10            10,812 Workflow.cs
```

Next, I remove the `OutputDir` and and the `GrpcOutputDir` options from the `Protobuf` item set definition, clean all previously generated files and build the project again. Now, everything succeeds. As expected, the generated CS files are now inside the designated `obj` folder. However, the previously missing `XxxGrpc.cs` files are now preset. One sibling file for each generated file that was generated before. Notably, all the `XxxGrpc.cs` files are empty. Here is an example output folder:

```
 Directory of c:\Repos\SDK.NET\_build\obj\Debug-AnyCPU\SDK\RawClient-Managed\net6.0\temporal\api\enums\v1

2022-03-18  15:39             4,306 CommandType.cs
2022-03-18  15:39                 0 CommandTypeGrpc.cs
2022-03-18  15:39             3,438 Common.cs
2022-03-18  15:39                 0 CommonGrpc.cs
2022-03-18  15:39            15,804 EventType.cs
2022-03-18  15:39                 0 EventTypeGrpc.cs
2022-03-18  15:39            11,366 FailedCause.cs
2022-03-18  15:39                 0 FailedCauseGrpc.cs
2022-03-18  15:39             3,193 Namespace.cs
2022-03-18  15:39                 0 NamespaceGrpc.cs
2022-03-18  15:39             3,153 Query.cs
2022-03-18  15:39                 0 QueryGrpc.cs
2022-03-18  15:39             2,130 Reset.cs
2022-03-18  15:39                 0 ResetGrpc.cs
2022-03-18  15:39             3,579 TaskQueue.cs
2022-03-18  15:39                 0 TaskQueueGrpc.cs
2022-03-18  15:39            10,812 Workflow.cs
2022-03-18  15:39                 0 WorkflowGrpc.cs
```

(Same for _netcoreapp3.1_ and for _net6.0_)

### Important note

Please note the location of the `obj` folder above. It has been placed outside its usual location in the project file by setting `BaseIntermediateOutputPath` and `IntermediateOutputPath`.

**Comment by tonydnewell** [comment]
@macrogreg is this still a problem? If so, can you provide a minimal example project that reproduces the problem? i.e. the csproj, the .proto files and all the other files needed and upload e.g. a github repository.

**Comment by jzabroski** [comment]
I think this is still a problem, or some variation of it is still a problem.  I occasionally get annoying errors that protodep "file is in use by another process", when the only things touching protodep files are apparently https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets

I think the actual issue happens whenever you glob a bunch of .proto files into a ProtoBuf item, where one or more of the .proto files import another one of the .proto files (e.g., Common.proto).  I suspect it has something to do with out DependencyInputs and DependencyOutputs is calculated, but I don't see the bug.

**Comment by jzabroski** [comment]
I believe this problem may be solvable using RemoveDuplicates "list operator" in MSBuild https://github.com/dotnet/msbuild/blob/main/src/Tasks/ListOperators/RemoveDuplicates.cs

Alternatively, ProtoReadDependencies MSBuild task can be amended to prevent duplicates in the first place. However, I am not really clear on what the purpose of cached protodep files are. I tried to find an explanation somewhere and it was remarkably unclear (nothing other than noting protodep files are a cache of proto file dependencies, which seems rather circular or not helpful to me).

 https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Tools/ProtoReadDependencies.cs#L58-L67

**Comment by tonydnewell** [comment]
@jzabroski Do you have an example that reproduces the problem? At least a directory layout, the proto files and which ones are imported by others?

**Comment by jzabroski** [comment]
I'm working on something reproducible but it doesn't happen every build. The particular .proto files are under IP restrictions where if I were to share them, a zero cure time service cancelation would trigger in the contract.

**Comment by tonydnewell** [comment]
@jzabroski The contents of the proto files shouldn't be important - just some files with minimal viable contents should be ok. I'm interested in the location of the files (relative to each other and the project) and dependencies on each other.

**Comment by jzabroski** [comment]
I'll get you a repro either today or tomorrow (Eastern Standard Time here). Have to first wrap up some basic tests today and make progress on the project.

But, as an observation, I can see @macrogreg also has a Common.proto file in his sample that is _likely_ imported using v3 proto syntax.  This is my first ever gRPC project, so I dont know when v3 proto syntax was introduced, but it seems like the Grpc.Tools are at least 5 years old judging by commit history; it is possible the tool chain was written before imports were possible. That is also the vibe I got from the Chromium c# build integration I read.

**Comment by tonydnewell** [comment]
@jzabroski just a ping to see if there is an update

**Comment by jzabroski** [comment]
I will work on providing a repro today. Thanks for the ping!

**Comment by jzabroski** [comment]
I have not been able to reproduce. It may be a build break due to MSBuild's FastUpToDateCheck.  In either case, I think I am right about the root cause.

One thing I see consistently in the build logs is:

```
WARNING: Potential incremental build failure in Simmaculate.MyProject.TaskRunner.csproj'. See: https://aka.ms/incremental-build-failure
```

The URL redirects to https://github.com/dotnet/project-system/blob/main/docs/up-to-date-check.md .

So, I suspect the issue is actually a mix of MSBuild doing a bad up to date check and the repeat items in the collection.  One thing I could try is forcing a duplicate in the task that calls ProtoCompile and see if the error matches; that would eliminate the difficulty in reproducing an I/O race condition.
```

**Comment by jzabroski** [comment]
So, I was able to reproduce the issue locally (using large solution, not small repro), and the actual behavior astounded me.  MSBuild is building this particular project 3 times - one build is for the rebuild, another build is for designtimebuild.  I captured a binlog, so I just need to investigate further to figure out what is happening.

But here is the human readable log - it may raise some awareness of things.  One thing that is confusing to me is _Protobuf_CoreClean is deleting the protodep, but then _Protobuf_GatherStaleBatched is skipped, and then _Protobuf_CoreCompile runs - this problem therefore likely only occurs on `rebuild` target, not `build` but I also think the build type plays a role.  The problem itself of design time builds causing issues was only visible to me by installing the project system tools plug-in to Visual Studio 2022; https://github.com/dotnet/project-system-tools#build-logging

33>Target _Protobuf_CoreClean:
33>    Deleting file "Protos\MyEntities.cs".
33>    Deleting file "obj\Debug\net48\12fbc7ffe642ca7e_MyEntities.protodep".
33>Target _GetProjectReferenceTargetFrameworkProperties:
33>Target ResolveProjectReferences:
33>Target GenerateTargetFrameworkMonikerAttribute:
33>  Skipping target "GenerateTargetFrameworkMonikerAttribute" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_GatherStaleBatched:
33>  Skipping target "_Protobuf_GatherStaleBatched" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_GatherStaleBatched:
33>  Skipping target "_Protobuf_GatherStaleBatched" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_GatherStaleBatched:
33>  Skipping target "_Protobuf_GatherStaleBatched" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_GatherStaleBatched:
33>  Skipping target "_Protobuf_GatherStaleBatched" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_GatherStaleBatched:
33>  Skipping target "_Protobuf_GatherStaleBatched" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_GatherStaleBatched:
33>  Skipping target "_Protobuf_GatherStaleBatched" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_GatherStaleBatched:
33>  Skipping target "_Protobuf_GatherStaleBatched" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_GatherStaleBatched:
33>  Skipping target "_Protobuf_GatherStaleBatched" because all output files are up-to-date with respect to the input files.
33>Target _Protobuf_CoreCompile:
33>    D:\Users\John.Zabroski\.nuget\packages\grpc.tools\2.50.0\build\_protobuf\Google.Protobuf.Tools.targets(280,5): error : Unable to load obj\Debug\net48\12fbc7ffe642ca7e_MyEntities.protodep: IOException: The process cannot access the file 'D:\Users\John.Zabroski\source\Simmaculate.MyProduct.MyProject\obj\Debug\net48\12fbc7ffe642ca7e_MyEntities.protodep' because it is being used by another process.
33>Done building target "_Protobuf_CoreCompile" in project "Simmaculate.MyProduct.MyProject.csproj" -- FAILED.
33>
33>Done building project "Simmaculate.MyProduct.MyProject.csproj" -- FAILED.
33>
33>Build FAILED.

**Comment by jzabroski** [comment]
@tonydnewell Got it.  Here is the issue: https://github.com/dotnet/project-system/blob/main/docs/design-time-builds.md#:~:text=Design%2Dtime%20builds%20are%20special%20builds%20that%20are%20launched%20by%20the%20project%20system%20to%20gather%20just%20enough%20information%20to%20populate%20the%20language%20service%20and%20other%20project%20services%2C%20such%20as%20the%20Dependencies%20node.

The protobuf sdk is not safe for designtime builds under certain conditions.

**Comment by jzabroski** [comment]
The workaround (which does not work if you have a CI build server with dirty/incremental checkouts and rebuild target instead of build target) is to build the 'bottom most' project in your dependency hierarchy that contains proto files by itself.  Or at least, that seems to be the workaround (in case anyone else finds this awful race condition problem)

**Comment by jzabroski** [comment]
BTW, this would also explain the errors "Some protoc outputs were not generated" for other users. I bet the missing outputs were either deleted or generated in design time mode.

**Comment by tonydnewell** [comment]
@jzabroski thank you for your input. It's on my list to look at.

**Comment by jzabroski** [comment]
I logged a bug with the project-system team, but if you read their docs closely, they may close it as a wont-fix since the docs imply that custom targets should be aware of designtimebuild bleeding into build.  I personally think both sides (project-system and Grpc.Tools) are buggy. Fixing both would be ideal.  The project-system fix is to not execute the same project concurrently, twice. Put another way, DesignTimeBuild should not execute concurrently with Build.

**Comment by tonydnewell** [comment]
@jzabroski I'm getting a better understanding of the build system, particularly the Grpc.Tools MSBuild targets.
I'm not sure I fully understand your build environment. Let me try and explain a few things about how the Grpc.Tools builds work (you may have already worked this out) and see if this helps us get further.

**Dependencies files:** The _.protodep_ files are written out by the the protocol buffers compiler so that we know which files are generated for each _.proto_ file.  There are two reasons for reading the dependency files:
- checking for incremental builds - the file from a previous build can be read to check the generated files against the _.proto_ files to see if they need to be recreated.
- finding which files were actually generated by the protocol buffers compiler so they can be added to the `<Compile>` items for the CS compiler

**Design-time builds:** These are builds done by Visual Studio when files are saved etc to see if any dependencies need updating without doing a full build.  For example, if you add/delete/edit a _.proto_ file then the generated _.cs_ files may need to be recreated otherwise you'd get errors in you other files if they refer to types that have changed.

By default for the Grpc.Tools targets, the design-time builds and normal builds are identical since we always want to make sure the generated files are up to date in Visual Studio.  You can disable calling the protocol buffers compiler in design time builds by setting _DisableProtobufDesignTimeBuild_ property to _true_ in your project file **if** it is a desgin time build, e.g. by adding
```xml
<PropertyGroup Condition="'$(DesignTimeBuild)' == 'true'
     OR '$(BuildingProject)' != 'true'">
    <DisableProtobufDesignTimeBuild>true</DisableProtobufDesignTimeBuild>
</PropertyGroup>
```

I'm not sure why you are seeing the file access errors when reading the dependencies files. This is happening just after the protocol buffers compiler is called to find out what files were generated.  The file should exist and nothing else should be running in parallel as far as I understand (MSBuild can run Projects in parallel, but not tasks as far as I'm away - I may be wrong).  If you have any more detailed logging that would be useful.  I shall continue to investigate.

**Comment by jzabroski** [comment]
Hi @tonydnewell  - I do have a build log, but I can't share it publicly. I'm not sure what protocols grpc and your employer (Google?) have in place on sharing somewhat confidential information. I'd also need to double check with my client that the are OK with me sharing it.

But, I would like to point out, the `DisableProtobufDesignTimeBuild` MSBuild property was added fairly recently, I think by a Microsoft intern, and the PR was not well explained as to what problem this was meant to solve.  I suspect it _was_ a bit of a kludge to prevent the concurrent build issue.  I suspect the _fix_ is to allow multiple processes to allow reading dependency files from within the protoc compiler (and therefore the MSBuild task).  This idea was also recommended to me by an MSBuild expert (Nirmal Guru, yes, his last name is Guru and he is a guru) as the most likely ideal resolution, but neither of us know enough about grpc project to say for sure.  I'm not saying we should remove `DisableProtobufDesignTimeBuild`, as I think it is probably useful in some cases.  I'm rather saying we should make it so that the default (not disabled) works without errors.

BTW, I want to thank you for such a candid reply and update.  Wish more people were like this. A+.

**Comment by tonydnewell** [comment]
Don't share anything that is private or confidential.

The error you are seeing is in the `ProtoCompile` custom task called from the `_Protobuf_CoreCompile` target.
I don't understand how it is happening since the task does:
- call the `protoc` executable to compile to ._proto_ file - this generates the ._cs_ files and writes out the ._protodep_ file
- then read the ._protodep_ file to find out which files that `protoc` generated

The second step fails. There shouldn't be anything else locking that file.
Maybe:
- the file doesn't get written? Have you checked it exists after the error?
- something else is rewriting the file - e.g. targets happening in parallel. If so, how? Multiple projects can be built in parallel but not targets within a project.

If you are able to extract and remove confidential info from a diagnostic log for just the `_Protobuf_CoreCompile` targets (for a successful or failing build) then that might help.

Also what is in your ._csproj_ file? Do you specify multiple `<Protobuf>` items?

**Comment by jzabroski** [comment]
I got another, different, build break in our CI server. There, we don't do clean checkouts but only add latest via git pull head. Canceling the build during restore seemed to permanently mess up the grpc targets somehow. I tried to figure out what went wrong, but just documenting it as part of this discussion to log the flakiness with these targets.

I will give csproj layout in separate reply.

**Comment by tonydnewell** [comment]
FYI I'm in the process of updating the documentation for Grpc.Tools.
There are implementation notes for maintainers here: https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Tools/implementation_notes.md
and I'm updating the end user documentation (yet to be reviewed): https://github.com/grpc/grpc/pull/32075 - [BUILD-INTEGRATION.md](https://github.com/grpc/grpc/blob/f4e5afc77a1c6311ffea8475c7388c355e3546af/src/csharp/BUILD-INTEGRATION.md)

I think the original issue here reported by @macrogreg and the issues reported by @jzabroski are different.  For both I need more information.

@macrogreg is this still an issue? I think it may be to do with you specifying `GrpcServices = "client"` on files that only contain message definitions and no service definitions. More details would help.

@jzabroski can you open a new issue and give your specific problems with some examples?

**Comment by jzabroski** [comment]
https://github.com/grpc/grpc/blob/f4e5afc77a1c6311ffea8475c7388c355e3546af/src/csharp/BUILD-INTEGRATION.md#bypassing-grpctools-to-run-the-protocol-buffers-compiler-explicitly - this is what I ended up doing, but my full target looks like:

```xml
  <Target Name="GenerateFlexTradeClasses">
    <PropertyGroup>
      <GoogleProtobufToolsVersion>3.21.12</GoogleProtobufToolsVersion>
      <GrpcToolsVersion>2.51.0</GrpcToolsVersion>
      <ProtoFileInputPath>$(SourceDirectory)\Simmaculate.Client\Protos</ProtoFileInputPath>
      <Protobuf_ProtocFullPath>$(MSBuildProjectDirectory)\tools\packages\Google.Protobuf.Tools.$(GoogleProtobufToolsVersion)\tools\windows_x64\protoc</Protobuf_ProtocFullPath>
      <ProtobufClassesOutputPath>$(ProtoFileInputPath)</ProtobufClassesOutputPath>
      <Protobuf_gRPC_PluginFullPath>$(MSBuildProjectDirectory)\tools\packages\Grpc.Tools.$(GrpcToolsVersion)\tools\windows_x64\grpc_csharp_plugin.exe</Protobuf_gRPC_PluginFullPath>
    </PropertyGroup>
    <ItemGroup>
      <ProtoFilesToCompile Include="$(ProtoFileInputPath)\**.proto" />
    </ItemGroup>
    
    <PropertyGroup>
      <ProtoCCommand>
        $(Protobuf_ProtocFullPath) --plugin=protoc-gen-grpc=$(Protobuf_gRPC_PluginFullPath) --grpc_out=$(ProtobufClassesOutputPath) --grpc_opt=no_server --csharp_out=$(ProtobufClassesOutputPath) --proto_path $(ProtoFileInputPath) @(ProtoFilesToCompile, ', ')
      </ProtoCCommand>
    </PropertyGroup>
    <Message Importance="high" Text="ProtoCCommand=$(ProtoCCommand)" />
    <Exec Command="$(ProtoCCommand)" />
  </Target>
```

and I have a custom InstallTools task that basically puts packages.config dependencies into a folder called tools - this is the old school, .NET Framework way of doing things.

Mainly sharing to note the old maintainer left a lot of details out on how to configure the command.

**Comment by jzabroski** [comment]
For what it's worth, I ultimately was able to whittle my build breaks down to including Grpc.Tools _at all_ in the build process.  I dont know if I will have time to share details, but it was eye opening to see some of the other errors I was running into that I have not even had time to report.  My impression is Grpc.Tools has many race conditions.

**Comment by tonydnewell** [comment]
I'm glad you've found a workaround.  I'm not sure that I understand why you have been seeing problems or what the race conditions are. Are you happy for this issue to be closed unless you are able to share some more details?

A few things I don't understand:
* why you are seeing race conditions? - the only parallelism in MSBuild is that you can have multiple projects run in parallel. Within a project the targets are sequential.  Are you running multiple projects in parallel all trying to compile the same proto files? Do you have more than one instance of MSBuild (or MSBuild and Visual Studio) running at the same time?
* why you are seeing issues with Grpc.Tools even if you have no proto files in a project?
* why you need both Grpc.Tools and Google.Protobuf.Tools? Grpc.Tools contains the protoc compiler.

Did you try moving all the proto files to a separate project and having a dependency on that project from whatever projects needs them?

**Comment by jzabroski** [comment]
I am going to try to answer those 3 questions soon.  It fell off my radar because my workaround has me in a happy place for now.

**Comment by jtattermusch** [comment]
Doesn't like like we can make much progress here (and also seems like the originally filed issue is not being discussed in this thread anymore), so I'm going to close.
